src/AnalysisEpic.cxx

```diff
@@ -0,0 +1,424 @@
+  TTreeReaderArray<Float_t> mcpart_psy(tr,       "MCParticles.momentum.y");
+  TTreeReaderArray<Float_t> mcpart_psz(tr,       "MCParticles.momentum.z");
+
+
+  // Reco tracks
+  TTreeReaderArray<Int_t> tracks_type(tr,  "ReconstructedChargedParticles.type"); // needs to be made an int eventually in actual EE code
+  TTreeReaderArray<Float_t> tracks_e(tr, "ReconstructedChargedParticles.energy");
+  TTreeReaderArray<Float_t> tracks_p_x(tr, "ReconstructedChargedParticles.momentum.x");
+  TTreeReaderArray<Float_t> tracks_p_y(tr, "ReconstructedChargedParticles.momentum.y");
+  TTreeReaderArray<Float_t> tracks_p_z(tr, "ReconstructedChargedParticles.momentum.z");
+  TTreeReaderArray<Int_t> tracks_PDG(tr,  "ReconstructedChargedParticles.PDG");
+  TTreeReaderArray<Float_t> tracks_CHI2PID(tr,  "ReconstructedChargedParticles.goodnessOfPID");
+  
+  // RecoAssociations
+  TTreeReaderArray<UInt_t> assoc_simID(tr, "ReconstructedChargedParticlesAssociations.simID");
+  TTreeReaderArray<UInt_t> assoc_recID(tr, "ReconstructedChargedParticlesAssociations.recID");
+  TTreeReaderArray<Float_t> assoc_weight(tr, "ReconstructedChargedParticlesAssociations.weight");
+
+  // calculate Q2 weights
+  CalculateEventQ2Weights();
+
+  // counters
+  Long64_t numNoBeam, numEle, numNoEle, numNoHadrons, numProxMatched;
+  numNoBeam = numEle = numNoEle = numNoHadrons = numProxMatched = 0;
+
+  
+
+  cout << "begin event loop..." << endl;
+  tr.SetEntriesRange(1,maxEvents);
+  do{
+    if(tr.GetCurrentEntry()%10000==0) cout << tr.GetCurrentEntry() << " events..." << endl;
+
+    // resets
+    kin->ResetHFS();
+    kinTrue->ResetHFS();
+
+    
+    double maxP = 0;
+    int genEleID = -1;
+    bool foundBeamElectron = false;
+    bool foundBeamIon = false;
+
+    // Index maps for particle sets
+    std::map <double,int> genidmap; // <pz, index_gen>
+    std::map <int,int> mcidmap; // <index_mc, index_gen>
+    std::map <int,int> trackidmap; // <index, index_mc>  
+    std::map <int,int> trackstatmap; // <index, genstatus>
+
+    // Particles vectors
+    // The index of the vectors correspond to their for loop idx
+    std::vector<Particles> genpart;    // mcID --> igen
+    std::vector<Particles> mcpart;     // mcID --> imc
+    std::vector<Particles> trackpart;  // mcID --> (imc of matching mcpart) or (-1 if no match is found)
+
+    /*
+      GenParticles loop
+    */
+
+    for(int igen=0; igen<hepmcp_PDG.GetSize(); igen++) {
+
+      int pid_ = hepmcp_PDG[igen];
+
+      double px_ = hepmcp_psx[igen];
+      double py_ = hepmcp_psy[igen];
+      double pz_ = hepmcp_psz[igen];
+      double e_  = hepmcp_E[igen];
+     
+      double p_ = sqrt(pow(hepmcp_psx[igen],2) + pow(hepmcp_psy[igen],2) + pow(hepmcp_psz[igen],2));
+      double mass_ = (fabs(pid_)==211)?constants::pimass:(fabs(pid_)==321)?constants::kmass:(fabs(pid_)==11)?constants::emass:(fabs(pid_)==13)?constants::mumass:(fabs(pid_)==2212)?constants::pmass:0.;
```

Comment:
Perhaps we can convert this to a `CommonConstants` method.

Reply:
It's done like this in `AnalysisEcce` too. Instead of improving it in this PR, let's open a new issue, since it's not the only common method that could be lifted to either `Analysis` or some common header.

Reply:
Unclear to me right now, but what is the disadvantage for calculating mass using the four momentum in GenParticles?

Reply:
Would that be equivalent to the true mass?